squirrel.frontier/src/main/java/org/dice_research/squirrel/data/uri/norm/NormalizerImpl.java

```diff
@@ -20,122 +24,218 @@
+    static {
+        defaultPortMap.put("http", 80);
+        defaultPortMap.put("https", 443);
+        defaultPortMap.put("ftp", 21);
+        defaultPortMap.put("ftps", 990);
+        defaultPortMap.put("sftp", 22);
+
+    }
+
+    /**
+     * list containing query parameters for session ids     *
+     */
+    private static final List<String> sessionIDs = new ArrayList<>();
+    static {
+        sessionIDs.add("sessionid");
+        sessionIDs.add("jsessionids");
+        sessionIDs.add("phpsessid");
+        sessionIDs.add("sid");
+    }
+
+    @Override
+    public CrawleableUri normalize(CrawleableUri uri) {
+        URI uriObject = uri.getUri();
+        boolean changed = false;
+        // normalize path
+        String path = uriObject.getRawPath();
+        if (path != null) {
+            String temp = normalizePath(path);
+            if (temp != path) {
+                path = temp;
+                changed = true;
+            }
+        }
+
+
+        // Copy Normalization from
+        // https://github.com/crawler-commons/crawler-commons/blob/master/src/main/java/crawlercommons/filters/basic/BasicURLNormalizer.java
+        // OR use URI.normalize()
+
+        // Check whether the query part of a URI has to be sorted
+        String query = uriObject.getQuery();
+        if(query != null){
+            if(query.length() > 0) {
+                String[] queryList = query.split("&");
+                Arrays.sort(queryList);
+                List<String> queries = new ArrayList<>(Arrays.asList(queryList));
+                List<String> toRemove = new ArrayList<>();
+                for(String queryParameter : queries){
+                    //removing session ids
+                    if(sessionIDs.contains(queryParameter.split("=")[0].toLowerCase())){
+                        toRemove.add(queryParameter);
+                    }
+                }
+                queries.removeAll(toRemove);
+                query = String.join("&", queries);
+                changed = true;
+            }
+            else{
+                query = null;
+                changed = true;
+            }
+        }
+
+        //Remove default ports
+        int port = uriObject.getPort();
+        String scheme = uriObject.getScheme();
+        if(port != -1){
+            if(defaultPortMap.containsKey(scheme)){
+                if(port == defaultPortMap.get(scheme)){
+                    port = -1;
+                    changed = true;
+                }
+            }
+        }
+        // Filter fragments (i.e., delete them)
+        String fragment = uriObject.getFragment();
+        if ((fragment != null) && (fragment.length() > 0)) {
+            changed = true;
+        }
+
+        // convert host and scheme to lower case
+        String host = uriObject.getHost();
+        if(!scheme.equals(scheme.toLowerCase()) || !host.equals(host.toLowerCase())){
+            scheme = scheme.toLowerCase();
+            host = host.toLowerCase();
```

Comment:
It would be better to store the temporary result of `toLowerCase` to be able to reuse it inside the `if` statement instead of executing it again.